imap_processing/lo/l0/lo_science.py

```diff
@@ -0,0 +1,147 @@
+
+    Parameters
+    ----------
+    dataset : xr.Dataset
+        Lo science counts from packets_to_dataset function.
+    attr_mgr : ImapCdfAttributes
+        CDF attribute manager for Lo L1A.
+
+    Returns
+    -------
+    dataset : xr.Dataset
+        Parsed and decompressed histogram data.
+    """
+    hist_bin = dataset.sci_cnt
+
+    # initialize the starting bit for the sections of data
+    section_start = 0
+    # for each field type in the histogram data
+    for field in hist_data_meta:
+        data_meta = hist_data_meta[field]
+        # for each histogram binary string decompress
+        # the data
+        decompressed_data = [
+            decompress(
+                bin_str, data_meta.bit_length, section_start, data_meta.section_length
+            )
+            for bin_str in hist_bin.values
+        ]
+
+        # add on the epoch length (equal to number of packets) to the
+        # field shape
+        data_shape = (len(hist_bin), data_meta.shape[0], data_meta.shape[1])
+
+        # get the dimension names from the CDF attr manager
+        dims = [
+            value
+            for key, value in attr_mgr.get_variable_attributes(field).items()
+            if "DEPEND" in key
+        ]
+        # reshape the decompressed data
+        shaped_data = np.array(decompressed_data, dtype=np.uint32).reshape(data_shape)
+        # add the data to the dataset
+        dataset[field] = xr.DataArray(
+            shaped_data, dims=dims, attrs=attr_mgr.get_variable_attributes(field)
+        )
+
+        # increment for the start of the next section
+        section_start += data_meta.section_length
+
+    return dataset
+
+
+def decompress(
+    bin_str: str, bits_per_index: int, section_start: int, section_length: int
+) -> list[int]:
+    """
+    Parse and decompress binary histogram data for Lo.
+
+    Parameters
+    ----------
+    bin_str : str
+        Binary string to decompress.
+    bits_per_index : int
+        Number of bits per index of the data section.
+    section_start : int
+        The start bit for the section of data.
+    section_length : int
+        The length of the section of data.
+
+    Returns
+    -------
+    decompressed_ints : list[int]
+        Decompressed integers for the data section.
+    """
+    # select the decompression method based on the bit length
+    # of the compressed data
+    if bits_per_index == 8:
+        decompress = Decompress.DECOMPRESS8TO16
+    elif bits_per_index == 12:
+        decompress = Decompress.DECOMPRESS12TO16
+    else:
+        raise ValueError(f"Invalid bits_per_index: {bits_per_index}")
+
+    # parse the binary, convert to integers, and decompress
+    decompressed_ints = [
+        decompress_int(
+            int(bin_str[i : i + bits_per_index], 2),
+            decompress,
+            DECOMPRESSION_TABLES,
+        )
+        for i in range(section_start, section_start + section_length, bits_per_index)
+    ]
```

Comment:
You don't need to do this now, but something to think about since you're in the refactor mode.

I think you could rework this to be:
1. Get an array of all your ints `raw_ints = [int(bin_str[i : i + bits_per_index], 2) for i in range(section_start, section_start + section_length, bits_per_index)]`
2. Then call `decompress_int(raw_ints, decompress, DECOMPRESSION_TABLES)` by passing in the full array.

The reason I suggest that is currently you've got the function call in the hot-loop and there is a lot of extra verification/validation going on there. I think you could then switch over to numpy broadcasting/array lookup and do all of the decompression in one call. (I think SWAPI or SWE has an example of this with `np.take`, or just `lookup_table[raw_ints]` to do fancy indexing on your lookup table array.